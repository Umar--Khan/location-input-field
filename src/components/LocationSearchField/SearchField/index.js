const SearchField = () => {
  return <input name='location-input' />;
};

export default SearchField;
