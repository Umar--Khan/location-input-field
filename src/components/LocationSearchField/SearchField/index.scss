.location-input {
  &__container {
    display: flex;
    flex-direction: column;
    position: relative;
    margin-top: $space-030;
  }
}
