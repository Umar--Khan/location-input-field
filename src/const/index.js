export const NO_RESULTS_FOUND = 'No results found';

export const PLACE_TYPES_LABELS = {
  A: 'Airport',
  D: 'District',
  C: 'City',
  S: 'Station',
};

export const PLACE_TYPES_COLORS = {
  A: 'maroon',
  D: 'green',
  C: 'blue',
  S: 'dark-grey',
};
