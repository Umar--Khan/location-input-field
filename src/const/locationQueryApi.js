export const locationQueryApi = search_term =>
  `https://www.rentalcars.com/FTSAutocomplete.do?solrIndex=fts_en&solrRows=6&solrTerm=${search_term}`;
