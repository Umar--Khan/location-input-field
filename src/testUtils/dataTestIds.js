export const page = {
  pageTitle: 'page-title',
  locationSearchFieldTitle: 'location-search-field-title',
};

export const tags = {
  tagLabel: 'tags-label',
};

export const suggestions = {
  suggestionItem: 'suggestion-item',
  suggestionItemCountry: 'suggestion-item-country',
  suggestionItemName: 'suggestion-item-name',
  container: 'suggestions-container',
};

export const searchField = {
  inputField: 'search-input-field',
  inputFieldLabel: 'search-input-field-label',
};

export const app = {
  appContainer: 'app-container',
};
