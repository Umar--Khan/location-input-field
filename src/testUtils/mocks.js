export const locationSuggestions = [
  {
    'country': 'Bulgaria',
    'lng': 23.7,
    'searchType': 'G',
    'alternative': ['BG'],
    'index': 1,
    'bookingId': 'city-355866',
    'placeType': 'C',
    'placeKey': '100948',
    'countryIso': 'bg',
    'locationId': '-1',
    'name': 'Kʼoseva',
    'ufi': -835331,
    'isPopular': false,
    'region': 'Sofia',
    'lang': 'en',
    'lat': 42.4,
  },
  {
    'country': 'Bulgaria',
    'lng': 25.15,
    'searchType': 'G',
    'alternative': ['BG'],
    'index': 2,
    'bookingId': 'city-355411',
    'placeType': 'C',
    'placeKey': '100946',
    'countryIso': 'bg',
    'locationId': '-1',
    'name': 'Kʼochevtsi',
    'ufi': -835200,
    'isPopular': false,
    'region': 'Kardzhali',
    'lang': 'en',
    'lat': 41.6833,
  },
  {
    'country': 'Bulgaria',
    'lng': 25.3,
    'searchType': 'G',
    'alternative': ['BG'],
    'index': 3,
    'bookingId': 'city-355867',
    'placeType': 'C',
    'placeKey': '100949',
    'countryIso': 'bg',
    'locationId': '-1',
    'name': 'Kʼosevo',
    'ufi': -835333,
    'isPopular': false,
    'region': 'Kardzhali',
    'lang': 'en',
    'lat': 41.6167,
  },
  {
    'country': 'Bulgaria',
    'lng': 26.25,
    'searchType': 'G',
    'alternative': ['BG'],
    'index': 4,
    'bookingId': 'city-355869',
    'placeType': 'C',
    'placeKey': '100950',
    'countryIso': 'bg',
    'locationId': '-1',
    'name': 'Kʼosevtsi',
    'ufi': -835335,
    'isPopular': false,
    'region': 'Targovishte',
    'lang': 'en',
    'lat': 43.1,
  },
  {
    'country': 'Bulgaria',
    'lng': 26.6833,
    'searchType': 'G',
    'alternative': ['BG'],
    'index': 5,
    'bookingId': 'city-355624',
    'placeType': 'C',
    'placeKey': '100947',
    'countryIso': 'bg',
    'locationId': '-1',
    'name': 'Kʼolmen',
    'ufi': -835239,
    'isPopular': false,
    'region': 'Shumen',
    'lang': 'en',
    'lat': 43.0333,
  },
  {
    'country': 'Georgia',
    'lng': 41.4436,
    'searchType': 'G',
    'alternative': ['GE'],
    'index': 6,
    'bookingId': 'city-876445',
    'placeType': 'C',
    'placeKey': '449544',
    'countryIso': 'ge',
    'locationId': '-1',
    'name': 'Kʼochʼara',
    'ufi': -2328015,
    'isPopular': false,
    'lang': 'en',
    'lat': 42.8575,
  },
];
