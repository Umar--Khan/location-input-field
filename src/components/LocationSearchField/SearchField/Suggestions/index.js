import SuggestionsItem from 'components/LocationSearchField/SearchField/Suggestions/SuggestionsItem';
import { NO_RESULTS_FOUND } from 'const';
import { suggestions } from 'testUtils/dataTestIds';

const Suggestions = ({ suggestionsResults, isActive }) => {
  const renderSuggestionsResults = () =>
    suggestionsResults.map(suggestion => {
      if (suggestion.name === NO_RESULTS_FOUND) {
        return (
          <li
            className='suggestions-list-item'
            id={`suggestions-list-item${suggestion.index}`}
            key={suggestion.index}
            aria-describedby='pick-up-locations-suggestions-no-results'
            data-testid={suggestions.suggestionItem}
          >
            {NO_RESULTS_FOUND}
          </li>
        );
      }

      return (
        <SuggestionsItem suggestion={suggestion} key={suggestion.placeKey} />
      );
    });

  return (
    <div
      className={`suggestions-list-container${isActive ? ' active' : ''}`}
      aria-hidden={!isActive}
      data-testid={suggestions.container}
    >
      <ul
        className='suggestions-list'
        aria-labelledby='pick-up-locations-suggestions'
        role='listbox'
      >
        {renderSuggestionsResults()}
      </ul>
    </div>
  );
};

export default Suggestions;
