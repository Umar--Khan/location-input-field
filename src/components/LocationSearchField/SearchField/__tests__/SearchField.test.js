import renderer from 'react-test-renderer';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { render, fireEvent, waitFor } from '@testing-library/react';

import SearchField from 'components/LocationSearchField/SearchField';
import { searchField, suggestions } from 'testUtils/dataTestIds';
import { locationQueryApi } from 'const/locationQueryApi';
import { locationSuggestions } from 'testUtils/mocks';

const mockAxios = new MockAdapter(axios);

describe('<SearchField />', () => {
  beforeEach(() => {
    mockAxios.reset();
  });

  afterAll(() => {
    mockAxios.restore();
  });

  it('Render snapshot correctly', () => {
    const component = renderer.create(<SearchField />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Input Field', () => {
    const { getByTestId } = render(<SearchField />);

    const inputField = getByTestId(searchField.inputField);
    const inputFieldLabel = getByTestId(searchField.inputFieldLabel);

    expect(inputField).toBeInTheDocument();
    expect(inputFieldLabel.textContent).toBe('Pick-up Location');
  });

  it('Suggestions should not be active in beginning', () => {
    const { getByTestId } = render(<SearchField />);

    const suggestionsContainerClass = getByTestId(suggestions.container)
      .className;

    expect(suggestionsContainerClass).not.toContain('active');
  });

  it('Change in input field should fire a fetch of suggestions', async () => {
    const { getByTestId, getAllByTestId } = render(<SearchField />);

    const inputField = getByTestId(searchField.inputField);

    const query = 'birmingham';

    fireEvent.change(inputField, {
      target: { value: query },
    });

    mockAxios
      .onGet(locationQueryApi(query))
      .reply(200, { results: { docs: locationSuggestions } });

    await waitFor(() => {
      expect(getAllByTestId(suggestions.suggestionItem)).toHaveLength(6);
      expect(getByTestId(suggestions.container).className).toContain('active');
    });
  });

  it('No suggestions if there is an error from API', async () => {
    const { getByTestId, queryByTestId } = render(<SearchField />);

    const inputField = getByTestId(searchField.inputField);

    const query = 'birmingham';

    fireEvent.change(inputField, {
      target: { value: query },
    });

    mockAxios.onGet(locationQueryApi(query)).networkErrorOnce();

    await waitFor(() => {
      expect(queryByTestId(suggestions.suggestionItem)).not.toBeInTheDocument();
      expect(getByTestId(suggestions.container).className).not.toContain(
        'active',
      );
    });
  });
});
