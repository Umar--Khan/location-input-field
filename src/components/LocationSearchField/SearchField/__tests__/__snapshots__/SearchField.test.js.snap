// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchField /> Render snapshot correctly 1`] = `
<div
  className="location-input"
>
  <label
    className="location-input__label"
    data-testid="search-input-field-label"
    htmlFor="pick-up-location"
  >
    Pick-up Location
  </label>
  <div
    className="location-input__container"
  >
    <input
      aria-invalid="false"
      aria-label="Pick Up Location"
      autoComplete="off"
      className="undefined input-field"
      data-testid="search-input-field"
      name="pick-up-location"
      onChange={[Function]}
      placeholder="city, airport, station, region, district..."
      value=""
    />
    <div
      aria-hidden={true}
      className="suggestions-list-container"
      data-testid="suggestions-container"
    >
      <ul
        aria-labelledby="pick-up-locations-suggestions"
        className="suggestions-list"
        role="listbox"
      />
    </div>
  </div>
</div>
`;
